bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -862,6 +863,11 @@ public synchronized void start() {
 
         ledgerStorage.start();
 
+        // if ledgerStorage is SortedLedgerStorage, then add SortedLedgerStorageListener
+        if (ledgerStorage instanceof SortedLedgerStorage){
+            ((SortedLedgerStorage) ledgerStorage).addSortedLedgerStorageListener(getSortedLedgerStorageListener());
```

Comment:
This is not very good, it is bad practice to use instanceof in this case. We should have a common interface

Reply:
@ArvinDevel any reason you can't make this a general storage listener?

Reply:
Because flush failure is only relative to SortedLedgerStorage, so I just changed the SortedLedgerStorage to avoid much changes. Do you think that adding a common listener interface to LedgerStorage is ok? @sijie @eolivelli

Reply:
@ArvinDevel an alternative:

- make a StateManager interface, move the state management in Bookie into an implementation of `StateManager`. The state manager is taking care of switching states of a bookie.
- pass the StateManager into the ledger storage. different ledger storage implementations use this state manager to decide when to switch the state.